parsec/core/gui/claim_user_widget.py

```diff
@@ -1,207 +1,660 @@
 from parsec.core.gui.trio_thread import JobResultError, ThreadSafeQtSignal
 from parsec.core.gui.desktop import get_default_device
 from parsec.core.gui.custom_dialogs import show_error, GreyedDialog, show_info
 from parsec.core.gui.lang import translate as _
-from parsec.core.gui.password_validation import PasswordStrengthWidget, get_password_strength
+from parsec.core.gui.password_validation import get_password_strength
 from parsec.core.gui.ui.claim_user_widget import Ui_ClaimUserWidget
+from parsec.core.gui.ui.claim_user_code_exchange_widget import Ui_ClaimUserCodeExchangeWidget
+from parsec.core.gui.ui.claim_user_provide_info_widget import Ui_ClaimUserProvideInfoWidget
+from parsec.core.gui.ui.claim_user_instructions_widget import Ui_ClaimUserInstructionsWidget
+from parsec.core.gui.ui.claim_user_finalize_widget import Ui_ClaimUserFinalizeWidget
 
 logger = get_logger()
 
 
-async def _do_claim_user(
-    config,
-    password: str,
-    password_check: str,
-    token: str,
-    user_id: str,
-    device_name: str,
-    organization_addr: BackendOrganizationClaimUserAddr,
-):
-    if password != password_check:
-        raise JobResultError("password-mismatch")
-    if len(password) < 8:
-        raise JobResultError("password-size")
-
-    try:
-        device_id = DeviceID(f"{user_id}@{device_name}")
-    except ValueError as exc:
-        raise JobResultError("bad-device_name") from exc
-
-    try:
-        device = await core_claim_user(
-            organization_addr=organization_addr.to_organization_addr(),
-            new_device_id=device_id,
-            token=token,
-            keepalive=config.backend_connection_keepalive,
-        )
+class Claimer:
+    class Step(IntEnum):
+        RetrieveInfo = 1
+        WaitPeer = 2
+        GetGreeterSas = 3
+        SignifyTrust = 4
+        GetClaimerSas = 5
+        WaitPeerTrust = 6
+        ClaimUser = 7
 
-    except InviteClaimBackendOfflineError as exc:
-        raise JobResultError("backend-offline", info=str(exc)) from exc
+    def __init__(self):
+        self.main_oob_send, self.main_oob_recv = trio.open_memory_channel(0)
+        self.job_oob_send, self.job_oob_recv = trio.open_memory_channel(0)
 
-    except InviteClaimError as exc:
-        if "Cannot retrieve invitation creator" in str(exc):
-            raise JobResultError("not_found", info=str(exc)) from exc
-        else:
-            raise JobResultError("refused-by-backend", info=str(exc)) from exc
+    async def run(self, addr):
+        try:
+            async with backend_invited_cmds_factory(addr=addr) as cmds:
+                r = await self.main_oob_recv.receive()
 
-    try:
-        save_device_with_password(config.config_dir, device, password)
+                assert r == self.Step.RetrieveInfo
+                try:
+                    initial_ctx = await claimer_retrieve_info(cmds)
+                    await self.job_oob_send.send((True, None, initial_ctx.claimer_email))
```

Comment:
Would be better to create a class to describe what is exchanged in the channel given the tuple starts to be big...